File: src/ARCtrl/JsonIO/Table/Templates.fs
namespace ARCtrl.Json

open Thoth.Json.Core

open ARCtrl
open ARCtrl.Process
open ARCtrl.Helper

module Templates =

    let encoder (templates: Template []) =
        templates 
        |> Array.map (Template.encoder)
        |> Encode.array

    let decoder =
        Decode.array Template.decoder
        
    let fromJsonString (jsonString: string) =
        try Decode.fromJsonString decoder jsonString with
        | exn         -> failwithf "Error. Given json string cannot be parsed to Templates map: %A" exn

    let toJsonString (spaces: int) (templates: Template []) =
        Encode.toJsonString spaces (encoder templates)

[<AutoOpen>]
module TemplateExtension =

    type Template with

        static member fromJsonString (jsonString: string) =
            try Decode.fromJsonString Template.decoder jsonString with
            | exn     -> failwithf "Error. Given json string cannot be parsed to Template: %A" exn

        static member toJsonString(?spaces: int) =
            fun (template:Template) ->
                Encode.toJsonString (Encode.defaultSpaces spaces) (Template.encoder template)

        member this.toJsonString(?spaces: int) =
            Template.toJsonString(?spaces=spaces) this

        static member fromCompressedJsonString (s: string) =
            try Decode.fromJsonString (Compression.decode Template.decoderCompressed) s with
            | e -> failwithf "Error. Unable to parse json string to ArcStudy: %s" e.Message

        static member toCompressedJsonString(?spaces) =
            fun (obj:Template) ->
                let spaces = defaultArg spaces 0
                Encode.toJsonString spaces (Compression.encode Template.encoderCompressed obj)

        member this.toCompressedJsonString(?spaces) =
            Template.toCompressedJsonString(?spaces=spaces) this

File: js/JsonIO/Table/Templates.js
import { map } from "../../fable_modules/fable-library-js.4.22.0/Array.js";
import { Template_encoderCompressed, Template_decoderCompressed, Template_decoder, Template_encoder } from "../../Json/Table/Templates.js";
import { array as array_2 } from "../../fable_modules/Thoth.Json.Core.0.4.0/Decode.fs.js";
import { toText, printf, toFail } from "../../fable_modules/fable-library-js.4.22.0/String.js";
import { fromString } from "../../fable_modules/Thoth.Json.JavaScript.0.3.0/Decode.fs.js";
import { toString } from "../../fable_modules/Thoth.Json.JavaScript.0.3.0/Encode.fs.js";
import { defaultSpaces } from "../../Json/Encode.js";
import { defaultArg, unwrap } from "../../fable_modules/fable-library-js.4.22.0/Option.js";
import { encode, decode } from "./Compression.js";

export function Templates_encoder(templates) {
    const values = map(Template_encoder, templates);
    return {
        Encode(helpers) {
            const arg = map((v) => v.Encode(helpers), values);
            return helpers.encodeArray(arg);
        },
    };
}

export const Templates_decoder = array_2(Template_decoder);

export function Templates_fromJsonString(jsonString) {
    try {
        const matchValue = fromString(Templates_decoder, jsonString);
        if (matchValue.tag === 1) {
            throw new Error(toText(printf("Error decoding string: %O"))(matchValue.fields[0]));
        }
        else {
            return matchValue.fields[0];
        }
    }
    catch (exn) {
        return toFail(printf("Error. Given json string cannot be parsed to Templates map: %A"))(exn);
    }
}

export function Templates_toJsonString(spaces, templates) {
    return toString(spaces, Templates_encoder(templates));
}

export function ARCtrl_Template__Template_fromJsonString_Static_Z721C83C5(jsonString) {
    try {
        const matchValue = fromString(Template_decoder, jsonString);
        if (matchValue.tag === 1) {
            throw new Error(toText(printf("Error decoding string: %O"))(matchValue.fields[0]));
        }
        else {
            return matchValue.fields[0];
        }
    }
    catch (exn) {
        return toFail(printf("Error. Given json string cannot be parsed to Template: %A"))(exn);
    }
}

export function ARCtrl_Template__Template_toJsonString_Static_71136F3F(spaces) {
    return (template) => toString(defaultSpaces(spaces), Template_encoder(template));
}

export function ARCtrl_Template__Template_toJsonString_71136F3F(this$, spaces) {
    return ARCtrl_Template__Template_toJsonString_Static_71136F3F(unwrap(spaces))(this$);
}

export function ARCtrl_Template__Template_fromCompressedJsonString_Static_Z721C83C5(s) {
    try {
        const matchValue = fromString(decode(Template_decoderCompressed), s);
        if (matchValue.tag === 1) {
            throw new Error(toText(printf("Error decoding string: %O"))(matchValue.fields[0]));
        }
        else {
            return matchValue.fields[0];
        }
    }
    catch (e_1) {
        const arg_1 = e_1.message;
        return toFail(printf("Error. Unable to parse json string to ArcStudy: %s"))(arg_1);
    }
}

export function ARCtrl_Template__Template_toCompressedJsonString_Static_71136F3F(spaces) {
    return (obj) => toString(defaultArg(spaces, 0), encode(Template_encoderCompressed, obj));
}

export function ARCtrl_Template__Template_toCompressedJsonString_71136F3F(this$, spaces) {
    return ARCtrl_Template__Template_toCompressedJsonString_Static_71136F3F(unwrap(spaces))(this$);
}

